trezor-keyring.js

```diff
@@ -0,0 +1,534 @@
+const { EventEmitter } = require('events');
+const ethUtil = require('ethereumjs-util');
+const HDKey = require('hdkey');
+const { TransactionFactory } = require('@ethereumjs/tx');
+const transformTypedData = require('trezor-connect/lib/plugins/ethereum/typedData');
+
+const hdPathString = `m/44'/60'/0'/0`;
+const SLIP0044TestnetPath = `m/44'/1'/0'/0`;
+
+const ALLOWED_HD_PATHS = {
+  [hdPathString]: true,
+  [SLIP0044TestnetPath]: true,
+};
+
+const keyringType = 'Trezor Hardware';
+const pathBase = 'm';
+const MAX_INDEX = 1000;
+const DELAY_BETWEEN_POPUPS = 1000;
+const TREZOR_CONNECT_MANIFEST = {
+  email: 'support@metamask.io',
+  appUrl: 'https://metamask.io',
+};
+
+function wait(ms) {
+  return new Promise((resolve) => setTimeout(resolve, ms));
+}
+
+/**
+ * @typedef {import('@ethereumjs/tx').TypedTransaction} TypedTransaction
+ * @typedef {InstanceType<import("ethereumjs-tx")>} OldEthJsTransaction
+ */
+
+/**
+ * Check if the given transaction is made with ethereumjs-tx or @ethereumjs/tx
+ *
+ * Transactions built with older versions of ethereumjs-tx have a
+ * getChainId method that newer versions do not.
+ * Older versions are mutable
+ * while newer versions default to being immutable.
+ * Expected shape and type
+ * of data for v, r and s differ (Buffer (old) vs BN (new)).
+ *
+ * @param {TypedTransaction | OldEthJsTransaction} tx
+ * @returns {tx is OldEthJsTransaction} Returns `true` if tx is an old-style ethereumjs-tx transaction.
+ */
+function isOldStyleEthereumjsTx(tx) {
+  return typeof tx.getChainId === 'function';
+}
+
+class TrezorKeyring extends EventEmitter {
+  constructor({ bridge } = {}) {
+    super();
+    this.type = keyringType;
+    this.accounts = [];
+    this.hdPath = hdPathString;
+    this.hdk = new HDKey();
+    this.page = 0;
+    this.perPage = 5;
+    this.unlockedAccount = 0;
+    this.paths = {};
+
+    if (!bridge) {
+      throw new Error('Bridge is a required dependency for the keyring');
+    }
+
+    this.bridge = bridge;
+  }
+
+  /**
+   * Performs any asynchronous initialisation
+   *
+   * @returns {Promise<void>}
+   */
+  init() {
+    return this.bridge.init();
+  }
+
+  /**
+   * Gets the model, if known.
+   * This may be `undefined` if the model hasn't been loaded yet.
+   *
+   * @returns {"T" | "1" | undefined}
+   */
+  getModel() {
+    return this.bridge.model;
+  }
+
+  /**
+   * Clears the iframe created by Trezor
+   *
+   * @returns {Promise<void>}
+   */
+  dispose() {
+    return this.bridge.dispose();
+  }
+
+  serialize() {
+    return Promise.resolve({
+      hdPath: this.hdPath,
+      accounts: this.accounts,
+      page: this.page,
+      paths: this.paths,
+      perPage: this.perPage,
+      unlockedAccount: this.unlockedAccount,
+    });
+  }
+
+  deserialize(opts = {}) {
+    this.hdPath = opts.hdPath || hdPathString;
```

Comment:
Can we add this? (Idea being that all serializable data is deserialized:
```js
    if(opts.hdk){
      this.setHdk(opts.hdk);
    }

    this.paths = opts.paths || {};
    this.unlockedAccount = opts.unlockedAccount || 0;
```

and this method

```js
  setHdk ({ publicKey, chainCode }) {
    this.hdk.publicKey = Buffer.from(publicKey, 'hex')
    this.hdk.chainCode = Buffer.from(chainCode, 'hex')
  }
```